File: C#/ThreadManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Sockets;
using System.Threading;

public class ThreadManager
{
    public readonly ConcurrentDictionary<string, (Thread Thread, CancellationTokenSource CTS)> Threads = new ConcurrentDictionary<string, (Thread, CancellationTokenSource)>();
    public List<string> ThreadDescriptions = new List<string>();

    public List<string> GetThreadsList()
    {
        ThreadDescriptions.Clear(); // Clear existing descriptions

        foreach (var kvp in Threads)
        {
            Debug.WriteLine("Found Thread.");
            var name = kvp.Key;
            var thread = kvp.Value.Thread;
            var isAlive = thread.IsAlive; // Determine if the thread is alive
            ThreadDescriptions.Add($"{name} - Alive: {isAlive}");
        }

        return ThreadDescriptions;
    }

    // Starts a new thread with a given name and action
    public bool AddThread(string name, Thread thread, CancellationTokenSource cts)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        Debug.WriteLine("Adding Existing Thread with CTS");
        return Threads.TryAdd(name, (thread, cts));
    }

    public bool AddThread(string name, Action action)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        Debug.WriteLine("Adding New Thread with private CTS");

        var cts = new CancellationTokenSource();
        ThreadStart threadStart = () => action.Invoke();
        var newThread = new Thread(threadStart) { Name = name };
        Threads.TryAdd(name, (newThread, cts));
        newThread.Start();

        return true;
    }

    public bool AddThread(string name, Action action, CancellationTokenSource cts)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        ThreadStart threadStart = () => action.Invoke();
        var thread = new Thread(threadStart) { Name = name };
        Threads.TryAdd(name, (thread, cts));
        thread.Start();
        return true;
    }

    public bool AddThread(string name, Action<CancellationToken> action, CancellationTokenSource cts)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        ThreadStart threadStart = () => action.Invoke(cts.Token);
        var thread = new Thread(threadStart) { Name = name };
        Threads.TryAdd(name, (thread, cts));
        thread.Start();
        return true;
    }

    // Adds a new thread that takes a client parameter and an action to perform
    public bool AddThread(string name, TcpClient client, Action<TcpClient, CancellationToken> action)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        Debug.WriteLine("Adding New Thread with client action");
        var cts = new CancellationTokenSource();
        ThreadStart threadStart = () => action.Invoke(client, cts.Token);
        var thread = new Thread(threadStart) { Name = name };
        Threads.TryAdd(name, (thread, cts));
        thread.Start();

        return true;
    }

    public bool AddThread(string name, object client, Action<object> action)
    {
        if (Threads.ContainsKey(name))
        {
            return false;
        }

        Debug.WriteLine("Adding New Thread with client action");
        var cts = new CancellationTokenSource();
        ThreadStart threadStart = () => action.Invoke(client);
        var thread = new Thread(threadStart) { Name = name };
        Threads.TryAdd(name, (thread, cts));
        thread.Start();

        return true;
    }

    // Attempts to stop a thread by name
    public void StopThread(string name)
    {
        if (Threads.ContainsKey(name))
        {
            var threadTuple = Threads[name];
            var targetThread = threadTuple.Thread;
            var cts = threadTuple.CTS;
            cts.Cancel(); // Signal cancellation to the thread

            // Optional: Wait a bit for the thread to finish
            if (!targetThread.Join(TimeSpan.FromSeconds(2)))
            {
                targetThread.Interrupt(); // Force the thread to stop
            }
            Threads.TryRemove(name, out var _);
        }
    }

    // Attempts to stop all threads
    public void StopAllThreads()
    {
        Debug.WriteLine("Stopping all threads.");
        foreach (var kvp in Threads.Keys)
        {
            Debug.WriteLine($"Stopping thread: {kvp}.");
            StopThread(kvp);
        }
    }
}

File: VB/ThreadManager.vb
Imports System.Collections.Concurrent
Imports System.Collections.Generic
Imports System.Net.Sockets
Imports System.Threading

Public Class ThreadManager

  Public ReadOnly threads As New ConcurrentDictionary(Of String, (Thread As Thread, CTS As CancellationTokenSource))()
  Public threadDescriptions As New List(Of String)()

  Public Function GetThreadsList() As List(Of String)
  	threadDescriptions.Clear() ' Clear existing descriptions
  
  	For Each kvp In threads
  		Debug.WriteLine("Found Thread.")
  		Dim name = kvp.Key
  		Dim thread = kvp.Value.Thread
  		Dim isAlive = thread.IsAlive ' Determine if the thread is alive
  		threadDescriptions.Add($"{name} - Alive: {isAlive}")
  	Next
  
  	' Setting the DataSource to Nothing first forces the control to refresh
  	Return threadDescriptions
  End Function


	' Starts a new thread with a given name and action
	Public Function AddThread(name As String, thread As Thread, cts As CancellationTokenSource) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Debug.WriteLine("Adding Existing Thread with CTS")
		Return threads.TryAdd(name, (thread, cts))
	End Function

	Public Function AddThread(name As String, action As Action) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Debug.WriteLine("Adding New Thread with private CTS")

		Dim cts As New CancellationTokenSource()
		Dim threadStart As ThreadStart = Sub() action.Invoke()
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}
		threads.TryAdd(name, (thread, cts))
		thread.Start()

		Return True
	End Function

	Public Function AddThread(name As String, action As Action, cts As CancellationTokenSource) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Dim threadStart As ThreadStart = Sub() action.Invoke()
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}
		threads.TryAdd(name, (thread, cts))
		thread.Start()
	End Function

	Public Function AddThread(name As String, action As Action(Of CancellationToken), cts As CancellationTokenSource) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Dim threadStart As ThreadStart = Sub() action.Invoke(cts.Token)
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}
		threads.TryAdd(name, (thread, cts))
		thread.Start()
	End Function

	Public Function AddThread(name As String, action As Action(Of CancellationToken)) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Debug.WriteLine("Adding New Thread with CT action")
		Dim cts As New CancellationTokenSource()
		Dim threadStart As ThreadStart = Sub() action.Invoke(cts.Token)
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}

		threads.TryAdd(name, (thread, cts))
		thread.Start()

		Return True
	End Function

	' Adds a new thread that takes a client parameter and an action to perform
	Public Function AddThread(name As String, client As TcpClient, action As Action(Of TcpClient, CancellationToken)) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Debug.WriteLine("Adding New Thread with client action")
		Dim cts As New CancellationTokenSource()
		Dim threadStart As ThreadStart = Sub() action.Invoke(client, cts.Token)
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}

		threads.TryAdd(name, (thread, cts))
		thread.Start()

		Return True
	End Function

	Public Function AddThread(name As String, client As Object, action As Action(Of Object)) As Boolean
		If threads.ContainsKey(name) Then
			Return False
		End If

		Debug.WriteLine("Adding New Thread with client action")
		Dim cts As New CancellationTokenSource()
		Dim threadStart As ThreadStart = Sub() action.Invoke(client)
		Dim thread As New Thread(threadStart) With {
			.Name = name
		}

		threads.TryAdd(name, (thread, cts))
		thread.Start()

		Return True
	End Function

	' Attempts to stop a thread by name
	Public Sub StopThread(name As String)
		If threads.ContainsKey(name) Then
			Dim threadTuple = threads(name)
			Dim targetThread = threadTuple.Thread
			Dim cts = threadTuple.CTS
			cts.Cancel() ' Signal cancellation to the thread

			Try
				' Optional: Wait a bit for the thread to finish
				If Not targetThread.Join(TimeSpan.FromSeconds(2)) Then
					targetThread.Interrupt() ' Force the thread to stop
				End If
				threads.TryRemove(name, Nothing)
			Catch ex As Exception
			
			End Try
		Else

		End If
	End Sub

	' Attempts to stop all threads
	Public Sub StopAllThreads()
		Debug.WriteLine("Stopping all threads.")
		For Each kvp In threads
			Debug.WriteLine($"Stopping thread: {kvp.Key}.")
			StopThread(kvp.Key)
		Next
	End Sub
                    
End Class
